sdks/python/apache_beam/runners/interactive/interactive_beam.py

```diff
@@ -0,0 +1,199 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Module of the current iBeam (interactive Beam) environment.
+
+The purpose of the module is to reduce the learning curve of iBeam users, 
+provide a single place for importing and add sugar syntax for all iBeam
+components. It gives users capability to manipulate existing environment for
+interactive beam, TODO(ningk) run interactive pipeline on selected runner as
+normal pipeline, create pipeline with interactive runner and visualize
+PCollections as bounded dataset.
+
+Note: iBeam works the same as normal Beam with DirectRunner when not in an
+interactively environment such as Jupyter lab or Jupyter Notebook. You can also
+run pipeline created by iBeam as normal Beam pipeline by run_pipeline() with
+desired runners.
+"""
+
+import importlib
+
+import apache_beam as beam
+from apache_beam.runners.interactive import interactive_runner
+
+_ibeam_env = None
+
+
+def watch(watchable):
+  """Watches a watchable so that iBeam can understand your pipeline.
+
+  If you write Beam pipeline in a notebook or __main__ module directly, since
+  __main__ module is always watched by default, you don't have to instruct
+  iBeam. However, if your Beam pipeline is defined in some module other than
+  __main__, e.g., inside a class function or a unit test, you can watch() the
+  scope to instruct iBeam to apply magic to your pipeline when running pipeline
+  interactively.
+
+    For example:
+
+    class Foo(object)
+      def build_pipeline(self):
+        p = create_pipeline()
+        init_pcoll = p |  'Init Create' >> beam.Create(range(10))
+        watch(locals())
+        return p
+    Foo().build_pipeline().run()
+
+    iBeam will cache init_pcoll for the first run. You can use:
+
+    visualize(init_pcoll)
+
+    To visualize data from init_pcoll once the pipeline is executed. And if you
+    make change to the original pipeline by adding:
+
+    squares = init_pcoll | 'Square' >> beam.Map(lambda x: x*x)
+
+    When you re-run the pipeline from the line you just added, squares will
+    use the init_pcoll data cached so you can have an interactive experience.
+
+  Currently the implementation mainly watches for PCollection variables defined
+  in user code. A watchable can be a dictionary of variable metadata such as
+  locals(), a str name of a module, a module object or an instance of a class.
+  The variable can come from any scope even local variables in a method of a
+  class defined in a module.
+
+    Below are all valid:
+
+    watch(__main__)  # if import __main__ is already invoked
+    watch('__main__')  # does not require invoking import __main__ beforehand
+    watch(self)  # inside a class
+    watch(SomeInstance())  # an instance of a class
+    watch(locals())  # inside a function, watching local variables within
+  """
+  current_env().watch(watchable)
+
+
+def create_pipeline(runner=None, options=None, argv=None):
```

Comment:
This is not really very different than beam.Pipeline(...). It might be better to avoid adding. Using existing constructs will have the advantage of existing documentation and user knowledge, and user's can copy contents out of notebooks without much change.

Reply:
I'll make the change to only use interactive runner (and by default the underlying runner with direct runner) for now.

The function basically creates a pipeline object that is interactive. Currently we could only provide interactive experience if the underlying runner is DirectRunner (which makes the code equivalent to `beam.Pipeline(interactive_runner.InteractiveRunner(), ...))`.
Yeah, you're right, this is confusing. I'll remove the runner arg from the signature. In the future, when more underlying runners are supported, we can add back the runner arg and set the runner in the InteractiveRunner()
as `return beam.Pipeline(interactive_runner.InteractiveRunner(underlying_runner=runner))`